File: erpnext/vehicles/report/vehicle_maintenance_schedule/vehicle_maintenance_schedule.js
// Copyright (c) 2016, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Vehicle Maintenance Schedule"] = {
	"filters": [
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			default: frappe.defaults.get_user_default("Company"),
			reqd: 1
		},
		{
			fieldname: "date_type",
			label: __("Date Type"),
			fieldtype: "Select",
			options: ["Follow Up Date", "Service Due Date"],
			default: "Follow Up Date",
			reqd: 1,
		},
		{
			fieldname:"from_date",
			label: __("From Date"),
			fieldtype: "Date",
			default: frappe.datetime.get_today(),
			reqd: 1
		},
		{
			fieldname:"to_date",
			label: __("To Date"),
			fieldtype: "Date",
			default: frappe.datetime.get_today(),
			reqd: 1
		},
		{
			fieldname:"project_template",
			label: __("Project Template"),
			fieldtype: "Link",
			options: "Project Template"
		},
		{
			fieldname: "project_template_category",
			label: __("Project Template Category"),
			fieldtype: "Link",
			options: "Project Template Category"
		},
		{
			fieldname: "customer",
			label: __("Customer"),
			fieldtype: "Link",
			options: "Customer",
			get_query: function() {
				return {
					query: "erpnext.controllers.queries.customer_query"
				};
			}
		},
		{
			fieldname: "customer_group",
			label: __("Customer Group"),
			fieldtype: "Link",
			options: "Customer Group"
		},
		{
			fieldname: "variant_of",
			label: __("Model Item Code"),
			fieldtype: "Link",
			options: "Item",
			get_query: function() {
				return {
					query: "erpnext.controllers.queries.item_query",
					filters: {"is_vehicle": 1, "has_variants": 1, "include_disabled": 1}
				};
			}
		},
		{
			fieldname: "item_code",
			label: __("Variant Item Code"),
			fieldtype: "Link",
			options: "Item",
			get_query: function() {
				var variant_of = frappe.query_report.get_filter_value('variant_of');
				var filters = {"is_vehicle": 1, "include_disabled": 1};
				if (variant_of) {
					filters['variant_of'] = variant_of;
				}
				return {
					query: "erpnext.controllers.queries.item_query",
					filters: filters
				};
			}
		},
		{
			fieldname: "item_group",
			label: __("Item Group"),
			fieldtype: "Link",
			options: "Item Group"
		},
		{
			fieldname: "opportunity",
			label: __("Opportunity"),
			fieldtype: "Link",
			options: "Opportunity"
		},
	],

	prepare_column: function (column) {
		if (column && column.is_communication) {
			column.custom_editor = frappe.query_reports["Vehicle Maintenance Schedule"].get_remarks_editor;
		}
	},

	get_remarks_editor: function (column, current_value, rowIndex) {
		const row_data = this.datatable.datamanager.getData(rowIndex);
		this.datatable.cellmanager.deactivateEditing(false);

		if (["Lost", "Closed"].includes(row_data.status)) {
			frappe.msgprint(__("Cannot submit communication because Opportunity is {0}", [row_data.status]));
			return;
		}

		const dialog = new frappe.ui.Dialog({
			title: __('Submit Communication'),
			doc: {},
			fields: [
				{
					fieldname: "remarks",
					fieldtype: "Small Text",
					label: __("Remarks"),
					reqd: 1
				},
				{
					fieldname: "action",
					fieldtype: "Select",
					label: __("Action"),
					reqd: 1,
					options: ["", "Submit Remarks Only", "Schedule Follow Up", "Mark As Lost", "Mark As Closed", "Create Appointment"],
					onchange: () => {
						let is_followup = dialog.get_value("action") == "Schedule Follow Up";
						let is_lost = dialog.get_value("action") == "Mark As Lost";

						dialog.set_df_property("schedule_date", "reqd", is_followup ? 1 : 0);
						dialog.set_df_property("follow_up_days", "reqd", is_followup ? 1 : 0);

						dialog.set_df_property("lost_reason", "reqd", is_lost ? 1 : 0);
					}
				},
				{fieldtype: "Section Break", depends_on: "eval:doc.action == 'Schedule Follow Up'"},
				{
					label : "Follow Up in Days",
					fieldname: "follow_up_days",
					fieldtype: "Int",
					onchange: () => {
						let today = frappe.datetime.nowdate();
						let contact_date = frappe.datetime.add_days(today, dialog.get_value('follow_up_days'));
						dialog.set_value('schedule_date', contact_date);
					}
				},
				{fieldtype: "Column Break"},
				{
					label : "Schedule Date",
					fieldname: "schedule_date",
					fieldtype: "Date",
					onchange: () => {
						var today = frappe.datetime.get_today();
						var schedule_date = dialog.get_value('schedule_date');
						dialog.doc.follow_up_days = frappe.datetime.get_diff(schedule_date, today);
						dialog.get_field('follow_up_days').refresh();
					}
				},
				{fieldtype: "Section Break", depends_on: "eval:doc.action == 'Mark As Lost'"},
				{
					fieldtype: "Table MultiSelect",
					label: __("Lost Reasons"),
					fieldname: "lost_reason",
					options: 'Lost Reason Detail',
				},
				{fieldtype: "Section Break"},
				{
					fieldname: "customer",
					fieldtype: "Link",
					options: "Customer",
					label: __("Customer"),
					read_only: 1,
					default: row_data.customer,
				},
				{
					fieldname: "contact_no",
					fieldtype: "Data",
					label: __("Contact No"),
					read_only: 1,
					default: row_data.contact_no,
				},
				{fieldtype: "Column Break"},
				{
					fieldname: "customer_name",
					fieldtype: "Data",
					label: __("Customer Name"),
					read_only: 1,
					default: row_data.customer_name,
				},
				{fieldtype: "Section Break"},
				{
					fieldname: "template_name",
					fieldtype: "Data",
					label: __("Maintenance Type"),
					read_only: 1,
					default: row_data.project_template_name,
				},
				{fieldtype: "Column Break"},
				{
					fieldname: "due_date",
					fieldtype: "Date",
					label: __("Due Date"),
					read_only: 1,
					default: row_data.due_date,
				},
				{fieldtype: "Section Break"},
				{
					fieldname: "vehicle",
					fieldtype: "Link",
					options: "Vehicle",
					label: __("Vehicle"),
					read_only: 1,
					default: row_data.vehicle,
				},
				{
					fieldname: "license_plate",
					fieldtype: "Data",
					label: __("Registration #"),
					read_only: 1,
					default: row_data.license_plate,
				},
				{
					fieldname: "delivery_date",
					fieldtype: "Date",
					label: __("Delivery Date"),
					read_only: 1,
					default: row_data.delivery_date,
				},
				{fieldtype: "Column Break"},
				{
					fieldname: "item_code",
					fieldtype: "Data",
					label: __("Variant Code"),
					read_only: 1,
					default: row_data.item_code,
				},
				{
					fieldname: "chassis_no",
					fieldtype: "Data",
					label: __("Chassis #"),
					read_only: 1,
					default: row_data.chassis_no,
				},
				{
					fieldname: "age",
					fieldtype: "Data",
					label: __("Age"),
					read_only: 1,
					default: row_data.age,
				},
			],
			primary_action: (dialog_data) => {
				dialog.hide();

				return frappe.call({
					method: "erpnext.vehicles.report.vehicle_maintenance_schedule.vehicle_maintenance_schedule.submit_communication_with_action",
					args: {
						remarks: dialog_data.remarks,
						action: dialog_data.action,
						follow_up_date: dialog_data.schedule_date,
						lost_reason: dialog_data.lost_reason,
						opportunity: row_data.opportunity,
						maintenance_schedule: row_data.maintenance_schedule,
						maintenance_schedule_row: row_data.maintenance_schedule_row,
						filters: frappe.query_report.get_filter_values(),
					},
					callback: function(r) {
						if (r.message && !r.exc) {
							if (r.message.updated_row) {
								frappe.query_report.datatable.datamanager.data[rowIndex] = r.message.updated_row;
								erpnext.utils.query_report_local_refresh();
							}

							if (r.message.appointment_doc) {
								frappe.model.sync(r.message.appointment_doc);
								frappe.set_route("Form", r.message.appointment_doc.doctype, r.message.appointment_doc.name);
							}
						}
					},
					error: function() {
						erpnext.utils.query_report_local_refresh();
					},
				});
			}
		});
		dialog.show();
	},

	formatter: function(value, row, column, data, default_formatter) {
		let style = {};

		if (column.is_communication == "communcation_date") {
			style['font-weight'] = 'bold';
		}

		if (["Lost", "Closed"].includes(data.status)) {
			style['color'] = "#858585";
		}

		if (column.fieldname == 'status') {
			style['font-weight'] = 'bold';
			if (value == "Open") {
				style['color'] = "orange";
			} else if (value == "To Follow Up") {
				style['color'] = "blue";
			} else if (value == "Replied") {
				style['color'] = "purple";
			} else if (value == "Converted") {
				style['color'] = "green";
			}
		}

		return default_formatter(value, row, column, data, {css: style});
	},

	onload: function () {
		frappe.model.with_doctype("Lost Reason Detail");
	}
};

File: erpnext/vehicles/report/vehicle_maintenance_schedule/vehicle_maintenance_schedule.py
# Copyright (c) 2013, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _
from frappe.utils import cint, getdate, today, add_days
from dateutil.relativedelta import relativedelta
from frappe.contacts.doctype.contact.contact import get_default_contact
import json


def execute(filters=None):
	return VehicleMaintenanceSchedule(filters).run()


class VehicleMaintenanceSchedule:
	def __init__(self, filters=None):
		self.filters = frappe._dict(filters or {})

		self.filters.from_date = getdate(self.filters.from_date or today())
		self.filters.to_date = getdate(self.filters.to_date or today())

		self.maintenance_reminder_days = cint(frappe.db.get_value("CRM Settings", None, "maintenance_opportunity_reminder_days"))
		self.filters.follow_up_from_date = add_days(self.filters.from_date, self.maintenance_reminder_days)
		self.filters.follow_up_to_date = add_days(self.filters.to_date, self.maintenance_reminder_days)

		self.communication_columns = 3

		if self.filters.from_date > self.filters.to_date:
			frappe.throw(_("From Date cannot be after To Date"))

	def run(self):
		self.get_data()
		self.get_communication_map()
		self.get_follow_up_map()
		self.process_data()
		self.get_columns()
		return self.columns, self.data

	def get_data(self):
		conditions = self.get_conditions()

		self.data = frappe.db.sql("""
			SELECT
				msd.scheduled_date as due_date, msd.project_template, msd.name as maintenance_schedule_row,
				ms.name as maintenance_schedule, ms.customer, ms.customer_name, ms.contact_mobile, ms.contact_phone,
				v.name as vehicle, v.item_code, v.delivery_date, v.chassis_no,
				v.engine_no, v.license_plate, v.unregistered, v.variant_of_name,
				v.customer as vehicle_customer, v.customer_name as vehicle_customer_name,
				pt.project_template_name, opp.name as opportunity, opp.status
			FROM `tabMaintenance Schedule Detail` msd
			LEFT JOIN `tabProject Template` pt ON pt.name = msd.project_template
			LEFT JOIN `tabMaintenance Schedule` ms ON ms.name = msd.parent
			LEFT JOIN `tabVehicle` v ON v.name = ms.serial_no
			LEFT JOIN `tabCustomer` c ON c.name = ms.customer
			LEFT JOIN `tabItem` im ON im.name = v.item_code
			LEFT JOIN `tabOpportunity` opp ON opp.maintenance_schedule = msd.parent AND opp.maintenance_schedule_row = msd.name
			WHERE ifnull(ms.serial_no, '') != '' AND {conditions}
		""".format(conditions=conditions), self.filters, as_dict=1)

	def get_conditions(self):
		conditions = []

		if self.filters.get("date_type") == "Follow Up Date":
			conditions.append("""(
				msd.scheduled_date BETWEEN %(follow_up_from_date)s AND %(follow_up_to_date)s
				OR exists(select fup.name from `tabLead Follow Up` fup where fup.parenttype = 'Opportunity'
					and fup.parent = opp.name and fup.schedule_date BETWEEN %(from_date)s AND %(to_date)s)
			)""")
		else:
			conditions.append("msd.scheduled_date BETWEEN %(from_date)s AND %(to_date)s")

		if self.filters.get("company"):
			conditions.append("ms.company = %(company)s")

		if self.filters.get("project_template"):
			conditions.append("msd.project_template = %(project_template)s")

		if self.filters.get("project_template_category"):
			conditions.append("pt.project_template_category = %(project_template_category)s")

		if self.filters.get("customer"):
			conditions.append("c.name = %(customer)s")

		if self.filters.get("customer_group"):
			lft, rgt = frappe.db.get_value("Customer Group", self.filters.customer_group, ["lft", "rgt"])
			conditions.append("""c.customer_group in (select name from `tabCustomer Group`
				where lft>=%s and rgt<=%s and docstatus<2)""" % (lft, rgt))

		if self.filters.get("variant_of"):
			conditions.append("im.variant_of = %(variant_of)s")

		if self.filters.get("item_code"):
			conditions.append("im.name = %(item_code)s")

		if self.filters.get("item_group"):
			lft, rgt = frappe.db.get_value("Item Group", self.filters.item_group, ["lft", "rgt"])
			conditions.append("""im.item_group in (select name from `tabItem Group`
				where lft>=%s and rgt<=%s and docstatus<2)""" % (lft, rgt))

		if self.filters.get("opportunity"):
			conditions.append("opp.name = %(opportunity)s")

		return " AND ".join(conditions) if conditions else ""

	def get_communication_map(self):
		self.communication_map = {}

		opportunities = [d.opportunity for d in self.data if d.get("opportunity")]
		if opportunities:
			communications = frappe.db.sql("""
				SELECT reference_name as opportunity, content,
					CAST(communication_date as DATE) as contact_date
				FROM tabCommunication
				WHERE reference_doctype = 'Opportunity' AND reference_name in %s
					AND communication_type != 'Automated Message' and sent_or_received = 'Received'
				ORDER BY communication_date, creation
			""", [opportunities], as_dict=1)

			for d in communications:
				self.communication_map.setdefault(d.opportunity, []).append(d)

	def get_follow_up_map(self):
		self.follow_up_map = {}

		opportunities = [d.opportunity for d in self.data if d.get("opportunity")]
		if opportunities:
			follow_ups = frappe.db.sql("""
				SELECT parent as opportunity, schedule_date
				FROM `tabLead Follow Up`
				WHERE parenttype = 'Opportunity' AND parent in %s AND ifnull(contact_date, '0000-00-00') = '0000-00-00'
				ORDER BY schedule_date, idx
			""", [opportunities], as_dict=1)

			for d in follow_ups:
				self.follow_up_map.setdefault(d.opportunity, []).append(d)

	def process_data(self):
		for d in self.data:
			d.disable_item_formatter = 1
			d.contact_no = d.contact_mobile or d.contact_phone

			if not d.variant_of_name:
				d.variant_of_name = d.item_name

			if not d.customer:
				d.customer = d.vehicle_customer
				d.customer_name = d.vehicle_customer_name

			if not d.contact_no:
				contact_id = get_default_contact('Customer', d.customer)
				d.contact_no = frappe.db.get_value("Contact", contact_id, "mobile_no", cache=1)

			d.age = self.get_formatted_duration(getdate(), d.delivery_date)
			d.reminder_date = add_days(d.due_date, -1 * self.maintenance_reminder_days)

			if not d.license_plate and d.unregistered:
				d.license_plate = 'Unreg'

			if d.opportunity:
				contact_list = []

				if self.communication_map.get(d.opportunity):
					communications = self.communication_map.get(d.opportunity)
					for c in communications:
						contact_list.append(frappe._dict({
							"communication": c.content,
							"communication_date": c.contact_date,
						}))

				if self.follow_up_map.get(d.opportunity):
					follow_ups = self.follow_up_map.get(d.opportunity)
					for c in follow_ups:
						contact_list.append(frappe._dict({
							"communication_date": c.schedule_date,
						}))

				for i, c in enumerate(contact_list):
					idx = str(i + 1)
					d["communication_" + idx] = c.communication
					d["communication_date_" + idx] = c.communication_date

				self.communication_columns = max(self.communication_columns, len(contact_list))

		self.data = sorted(self.data, key=lambda d: (getdate(d.due_date), getdate(d.delivery_date)))

	def get_formatted_duration(self, start_date, end_date):
		delta = relativedelta(getdate(start_date), getdate(end_date))

		template = ['Y', 'M', 'D']
		data = [delta.years, delta.months, delta.days]

		duration = " ".join([str(x) + y for x, y in zip(data, template) if x or y == 'D'])
		if duration == '0D':
			duration = '-'

		return duration

	def get_columns(self):
		columns = [
			{
				"label": _("Due Date"),
				"fieldname": "due_date",
				"fieldtype": "Date",
				"width": 80
			},
			{
				"label": _("Reminder Date"),
				"fieldname": "reminder_date",
				"fieldtype": "Date",
				"width": 90
			},
			# {
			# 	"label": _("Model"),
			# 	"fieldname": "variant_of_name",
			# 	"fieldtype": "Data",
			# 	"width": 120
			# },
			{
				"label": _("Variant Code"),
				"fieldname": "item_code",
				"fieldtype": "Link",
				"options": "Item",
				"width": 120
			},
			# {
			# 	"label": _("Engine No"),
			# 	"fieldname": "engine_no",
			# 	"fieldtype": "Data",
			# 	"width": 115
			# },
			{
				"label": _("Template"),
				"fieldname": "project_template",
				"fieldtype": "Link",
				"options": "Project Template",
				"width": 100
			},
			# {
			# 	"label": _("Template Name"),
			# 	"fieldname": "project_template_name",
			# 	"fieldtype": "Data",
			# 	"width": 180
			# },
			{
				"label": _("Delivery Date"),
				"fieldname": "delivery_date",
				"fieldtype": "Date",
				"width": 90
			},
			{
				"label": _("Age"),
				"fieldname": "age",
				"fieldtype": "Data",
				"width": 80
			},
			{
				"label": _("Customer"),
				"fieldname": "customer",
				"fieldtype": "Link",
				"options": "Customer",
				"width": 80
			},
			{
				"label": _("Customer Name"),
				"fieldname": "customer_name",
				"fieldtype": "Data",
				"width": 150
			},
			{
				"label": _("Contact No"),
				"fieldname": "contact_no",
				"fieldtype": "Data",
				"width": 100
			},
			{
				"label": _("Status"),
				"fieldname": "status",
				"fieldtype": "Data",
				"width": 90
			},
		]

		for i in range(self.communication_columns):
			idx = str(i + 1)
			columns += [
				{
					"label": _("Contact {0}").format(idx),
					"fieldname": "communication_date_" + idx,
					"fieldtype": "Date",
					"width": 80,
					"is_communication": "communcation_date",
				},
				{
					"label": _("Communication {0}").format(idx),
					"fieldname": "communication_" + idx,
					"fieldtype": "Data",
					"options": "Opportunity",
					"width": 200,
					"editable": 1,
					"is_communication": "remarks",
				},
			]

		columns += [
			{
				"label": _("Vehicle"),
				"fieldname": "vehicle",
				"fieldtype": "Link",
				"options": "Vehicle",
				"width": 70
			},
			{
				"label": _("Reg No"),
				"fieldname": "license_plate",
				"fieldtype": "Data",
				"width": 80
			},
			{
				"label": _("Chassis No"),
				"fieldname": "chassis_no",
				"fieldtype": "Data",
				"width": 150
			},
			{
				"label": _("Schedule"),
				"fieldname": "maintenance_schedule",
				"fieldtype": "Link",
				"options": "Maintenance Schedule",
				"width": 80
			},
			{
				"label": _("Opportunity"),
				"fieldname": "opportunity",
				"fieldtype": "Link",
				"options": "Opportunity",
				"width": 80
			},
		]

		if self.filters.date_type != "Follow Up Date":
			columns = [d for d in columns if d.get('fieldname') != 'reminder_date']

		self.columns = columns


@frappe.whitelist()
def submit_communication_with_action(remarks, action, opportunity=None, follow_up_date=None, lost_reason=None,
		maintenance_schedule=None, maintenance_schedule_row=None, filters=None):
	from erpnext.crm.doctype.opportunity.opportunity import submit_communication_with_action

	out = submit_communication_with_action(remarks, action,
		opportunity=opportunity, follow_up_date=follow_up_date, lost_reason=lost_reason,
		maintenance_schedule=maintenance_schedule, maintenance_schedule_row=maintenance_schedule_row)

	if filters is not None:
		if isinstance(filters, str):
			filters = json.loads(filters)

		filters['opportunity'] = out.opportunity
		out['updated_row'] = execute(filters)
		out['updated_row'] = out['updated_row'][1][0]

	return out
